app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.java

```diff
@@ -65,18 +82,101 @@ public boolean onOptionsItemSelected(MenuItem item) {
                 final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) findPreference(getString(R.string.pref_key_manual_add_search_engine));
                 final boolean engineValid = pref.validateEngineNameAndShowError(engineName);
                 final boolean searchValid = pref.validateSearchQueryAndShowError(searchQuery);
+                final boolean isPartialSuccess = engineValid && searchValid;
 
-                final boolean isSuccess = engineValid && searchValid;
-                if (isSuccess) {
-                    final SharedPreferences sharedPreferences = getSearchEngineSharedPreferences();
-                    SearchEngineManager.addSearchEngine(sharedPreferences, getActivity(), engineName, searchQuery);
-                    Snackbar.make(rootView, R.string.search_add_confirmation, Snackbar.LENGTH_SHORT).show();
-                    getFragmentManager().popBackStack();
+                if (isPartialSuccess) {
+                    // TODO: add spinner, disable button until async task completes.
+                    // Hide the keyboard because:
+                    // - It's awkward to show the keyboard while waiting for a response
+                    // - We want it hidden when we return to the previous screen (on success)
+                    // - An expanded keyboard hides the success snackbar
+                    ViewUtils.hideKeyboard(rootView);
+                    new ValidateSearchEngineAsyncTask(this, engineName, searchQuery).execute();
+                } else {
+                    TelemetryWrapper.saveCustomSearchEngineEvent(false);
                 }
-                TelemetryWrapper.saveCustomSearchEngineEvent(isSuccess);
                 return true;
+
             default:
                 return super.onOptionsItemSelected(item);
         }
     }
+
+    private static class ValidateSearchEngineAsyncTask extends AsyncTask<Void, Void, Boolean> {
+        private final WeakReference<ManualAddSearchEngineSettingsFragment> thatWeakReference;
+        private final String engineName;
+        private final String query;
+
+        private ValidateSearchEngineAsyncTask(final ManualAddSearchEngineSettingsFragment that, final String engineName,
+                final String query) {
+            this.thatWeakReference = new WeakReference<>(that);
+            this.engineName = engineName;
+            this.query = query;
+        }
+
+        @Override
+        protected Boolean doInBackground(final Void... voids) {
+            final boolean isValidSearchQuery = isValidSearchQueryURL(query);
+            TelemetryWrapper.saveCustomSearchEngineEvent(isValidSearchQuery);
+            return isValidSearchQuery;
+        }
+
+        @Override
+        protected void onPostExecute(final Boolean isValidSearchQuery) {
+            super.onPostExecute(isValidSearchQuery);
+
+            final ManualAddSearchEngineSettingsFragment that = thatWeakReference.get();
+            if (that == null) {
```

Comment:
Oh I see, you handle the cancelled state here. Are Fragments always gc-ed quickly after they're destroyed? Might be better to also explicitly setCancelled in the Fragment's onDestroy.

Reply:
fwiw, onDestroy doesn't work – e.g. if the home button is pressed, the fragment is not destroyed. Note that FragmentManager.popBackStack (which we call after the async task finishes) throws if it's called when the app is in the background.